### PR TITLE
Added documentation for the First Fit Decreasing packing algo

### LINKED ISSUE #1098: Create web documentation for the First Fit Decreasing packing algorithm

### PR DESCRIPTION
**Pull request by avflor**
Fixes  #1098

### CODE REVIEW

File: website/content/docs/developers/packing/ffdpacking.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+Parameter 1 determines the RAM requirement of each component in the topology.
+If the requirement is not specified then a default value of 1GB is used. The First Fit Decreasing algorithm
+guarantees that the amount of memory allocated to a component is either the one specified by the user
+or the default one.
+
+The parameters 2-4 determine the maximum container size with respect to RAM, cpu cores and disk. 
+If one of these parameters is not specified by the user then
+the hint for the corresponding maximum container resource is set to the default resource requirement of 4 Heron instances. 
+
+Note that these values take into account only the resources allocated for the user's instances. Additional per container resources for system-related processes such as the stream manager can be added to the maximum container size defined above. Thus, the algorithm might eventually produce containers whose size is slighltly bigger that the boundary determined by parameters 2-4.
```
Comment (kramasamy) [inline website/content/docs/developers/packing/ffdpacking.md]: it should "slightly bigger than"

File: website/content/docs/developers/packing/ffdpacking.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
+---
+title: The First Fit Decreasing Packing Algorithm
+---
+
+### Overview
+
+This guide provides basic steps at using and tuning the First Fit Decreasing packing algorithm in order to  utilize the resources efficiently. This packing algorithm aims at utilizing as few containers as possible, thus limiting the overall resources used. The algorithm is based on the First Fit Decreasing heuristic for the binpacking problem. To enable the First Fit Decreasing algorithm, update the `packing.yaml` file as follows:
+
```
Comment (kramasamy) [inline website/content/docs/developers/packing/ffdpacking.md]: Can you describe a few scenarios when this packing is useful?

Comment (billonahill) [inline website/content/docs/developers/packing/ffdpacking.md]: also at a high level how the algo works.

File: website/content/docs/developers/packing/ffdpacking.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+# packing algorithm for packing instances into containers
+heron.class.packing.algorithm:    com.twitter.heron.packing.binpacking.FirstFitDecreasingPacking
+```
+
+The algorithm accepts as input the values of the following parameters:
+
+1. Component RAM 
+2. Hint for the maximum container RAM (`com.twitter.heron.api.Config.TOPOLOGY_CONTAINER_MAX_RAM_HINT`)
+3. Hint for the maximum Container CPU (`com.twitter.heron.api.Config.TOPOLOGY_CONTAINER_MAX_CPU_HINT`)
+4. Hint for the maximum Container Disk (`com.twitter.heron.api.Config.TOPOLOGY_CONTAINER_MAX_DISK_HINT`)
```
Comment (billonahill) [inline website/content/docs/developers/packing/ffdpacking.md]: In #3 and #4 Container is capitalized but other occurrences in the doc are now. We should be consistent.

File: website/content/docs/developers/packing/ffdpacking.md (hunk 1 above)
Comment (billonahill) [inline website/content/docs/developers/packing/ffdpacking.md]: should clarify if the padding is done automatically by the system or if the user controls this with settings.

also please line break at 100 chars to make commenting and IDE reading easier.

File: website/content/docs/developers/packing/ffdpacking.md [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+If one of these parameters is not specified by the user then
+the hint for the corresponding maximum container resource is set to the default resource requirement of 4 Heron instances. 
+
+Note that these values take into account only the resources allocated for the user's instances. Additional per container resources for system-related processes such as the stream manager can be added to the maximum container size defined above. Thus, the algorithm might eventually produce containers whose size is slighltly bigger that the boundary determined by parameters 2-4.
+The amount of the additional resources allocated to each container to account for additional internal Heron resource requirements, is determined by the padding percentage
+specified in parameter 5. The default value for this parameter is 10, which means that after a container has been filled with user instances, an additional 10% of resources will be allocated to it.
+
+Based on these parameters, the algorithm decides how to place the instances in the containers
+and also how many containers to use. The final result can contain heterogeneous containers. Note that this algorithm does not require the number of containers
```
Comment (billonahill) [inline website/content/docs/developers/packing/ffdpacking.md]: omit: also

File: website/content/docs/developers/packing/ffdpacking.md (hunk 1 above)
Comment (billonahill) [inline website/content/docs/developers/packing/ffdpacking.md]: just a nit, but please line break at 100 chars to make commenting and reading in an IDE easier.

File: website/content/docs/developers/packing/ffdpacking.md [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+The amount of the additional resources allocated to each container to account for additional internal Heron resource requirements, is determined by the padding percentage
+specified in parameter 5. The default value for this parameter is 10, which means that after a container has been filled with user instances, an additional 10% of resources will be allocated to it.
+
+Based on these parameters, the algorithm decides how to place the instances in the containers
+and also how many containers to use. The final result can contain heterogeneous containers. Note that this algorithm does not require the number of containers 
+as input. 
+
+### Configuring the First Fit Decreasing Packing Algorithm
+
+1. The methods `com.twitter.heron.api.Config.setContainerMaxRamHint(long bytes)`, `com.twitter.heron.api.Config.setContainerMaxCpuHint(float ncpus)`, `com.twitter.heron.api.Config.setContainerMaxDiskHint(long bytes)`
```
Comment (billonahill) [inline website/content/docs/developers/packing/ffdpacking.md]: We should also link to the FirstFitDecreasingPacking javadocs for a detailed explanation of the settings and methods.

**Comment by avflor** [comment]
@kramasamy @billonahill I think I have addressed all your comments. Please let me know how the doc looks like now.

**Comment by kramasamy** [comment]
@avflor - looks good to me. @billonahill - let us know if this is ok to go.

**Comment by billonahill** [comment]
👍

**Pull request merged**